File: Java/Sphere.java
// File: Sphere.java

// Author: Christopher Camargo
// CMPS 112: Winter 2014

public class Sphere {
	private Vec3 center; // Center of a sphere in 3D.
	private float radius, squaredRadius; // Radius information.
	private Vec3 ambient, diffuse, specular; // Material information.
	private float shininess; // Shine for specular term.

	// Main constructor for a Sphere.
	// Sphere(center x, center y, center z, radius, diffuse r, diffuse g, diffuse b, ambient, specular, shine)
	public Sphere(float c1, float c2, float c3, float r, float d1, float d2, float d3, float a, float s, float sh) {
		center = new Vec3(c1, c2, c3);
		radius = r;
		squaredRadius = r * r;

		ambient = new Vec3(a);
		diffuse = new Vec3(d1, d2, d3);
		specular = new Vec3(s);
		shininess = sh;
	}

	// Returns the normal of sphere at point p.
	public Vec3 normal(Vec3 p) {
		return (p.sub(center)).normalize();
	}

	// Tests if a ray intersects an instance of this Sphere.
	public boolean intersect(Ray ray) {
		Vec3 l = center.sub(ray.origin());

		float tca = l.dot(ray.direction());

		if (tca < 0)
			return false;

		float d2 = l.dot(l) - tca * tca;

		if (d2 > squaredRadius)
			return false;

		float thc = (float) Math.sqrt(squaredRadius - d2);

		ray.setT(tca - thc);

		return true;
	}

	// Returns ambient light of sphere.
	public Vec3 ambient() {
		return ambient;
	}

	// Returns diffuse light of sphere.
	public Vec3 diffuse() {
		return diffuse;
	}

	// Returns specular light of sphere.
	public Vec3 specular() {
		return specular;
	}
}

File: Haskell/Sphere.hs
-- File: Sphere.hs

-- Author: Christopher Camargo
-- CMPS 112: Winter 2014

-- Sphere module declaration.
module Sphere
( Sphere (Sphere)
, center
, radius
, diff
, ambient
, spec
, shine
, squaredRad
, isectBool
, getSphere
, getVal
, normal
, le
, tca
, d2
, thc
, intersect
) where

import Vec3
import Ray
import Data.List (find)

-- Sphere with center and radius. Also includes color data.
data Sphere = Sphere { center :: Vec3, radius :: Float, diff :: Vec3, ambient :: Vec3, spec :: Vec3, shine :: Float } deriving (Show, Eq)

squaredRad :: Sphere -> Float -- Returns the squared radius of a sphere.
squaredRad sphere = (radius sphere) * (radius sphere)

isectBool :: Int -> Int -> Sphere -> Bool -- Returns true if there is an intersection, false otherwise.
isectBool x y s = if (intersect (getRay x y) s) > 0.0 then True else False

getSphere :: Int -> Int -> [Sphere] -> Maybe Sphere -- Returns a sphere if it intersects a ray.
getSphere x y s = find (isectBool x y) s

getVal :: Int -> Int -> Sphere -> Float -- Returns the value of the rays parameter at an intersection.
getVal x y s = (intersect (getRay x y) s)

normal :: Sphere -> Vec3 -> Vec3 -- Returns the normal of a sphere at a point.
normal sphere point = normalize $ sub point (center sphere)

le :: Ray -> Sphere -> Vec3 -- Helper returns vector from ray origin to sphere center.
le r s = sub (center s) (origin r)

tca :: Ray -> Sphere -> Float -- Helper calculates distance from ray origin to sphere center.
tca r s = dot (le r s) (direction r)

d2 :: Ray -> Sphere -> Float -- Helper calculates distance from sphere center to ray origin.
d2 r s = dot (le r s) (le r s) - (tca r s) * (tca r s)

thc :: Ray -> Sphere -> Float -- Helper calculates distance from ray intersection point on sphere to center.
thc r s = sqrt (squaredRad s - (d2 r s))

intersect :: Ray -> Sphere -> Float -- Returns parameter t > 0 if ray intersects sphere.
intersect r s = if (tca r s) < 0 
				then 0.0 
				else if (d2 r s) > squaredRad s 
					 then 0.0 
					 else (tca r s) - (thc r s)